### PR TITLE
Partial implementation of virtual ports for PyProcModels

### PR DESCRIPTION
**Pull request by mathisrichter**
Issue Number: #185 (partial fulfilment)

Objective of pull request: Supplies an implementation of virtual ports (reshape, flatten, transpose) for PyProcModels

## Pull request checklist

Your PR fulfills the following requirements:
- [x] [Issue](https://github.com/lava-nc/lava/issues) created that explains the change and why it's needed
- [x] Tests are part of the PR (for bug fixes / features)
- [ ] [Docs](https://github.com/lava-nc/docs) reviewed and added / updated if needed (for bug fixes / features)
- [x] PR conforms to [Coding Conventions](https://lava-nc.org/developer_guide.html#coding-conventions)
- [x] [PR applys BSD 3-clause or LGPL2.1+ Licenses](https://lava-nc.org/developer_guide.html#add-a-license) to all code files
- [x] Lint (`pyb`) passes locally
- [x] Build tests (`pyb -E unit`) or (`python -m unittest`) passes locally

## Pull request type

 

 

Please check your PR type:
- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation changes
- [ ] Other (please describe): 

## What is the current behavior?

- Virtual ports cannot be executed

## What is the new behavior?

- Virtual ports reshape, flatten, transpose can be executed with PyProcModels

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

**Review by PhilippPlank — COMMENTED** [review]
I think this looks good so far. When the OutPorts are also added, we should pull some of the loops looking for virtual ports in the compiler into a sub method.

File: src/lava/magma/core/model/py/ports.py [hunk 1]
```diff
@@ -61,18 +66,37 @@ def probe(self) -> bool:
...
+        ----------
+        recv_data : numpy.ndarray
+            data received on the port that shall be transformed
+
+        Returns
+        -------
+        recv_data : numpy.ndarray
+            received data, transformed by the incoming virtual ports
+        """
+        if self._transform_funcs is not None:
```
[the remark this answers is not in the document]
Comment (mathisrichter) [reply src/lava/magma/core/model/py/ports.py:209]: Yes, I'll change it

File: src/lava/magma/core/process/ports/ports.py [hunk 2]
```diff
@@ -1,8 +1,12 @@
 # Copyright (C) 2021 Intel Corporation
```
[the remark this answers is not in the document]
Comment (mathisrichter) [reply src/lava/magma/core/process/ports/ports.py:1]: True, I created an issue for that
https://github.com/lava-nc/lava/issues/188

**Comment by bamsumit** [comment]
Looks good. Can confirm transpose, reshape and flatten work with hierarchical process.

File: src/lava/magma/compiler/compiler.py [hunk 3]
```diff
@@ -344,12 +345,25 @@ def _compile_proc_models(
...
+                    ports = []
+                    for pt in (list(p.in_ports) + list(p.out_ports)):
+                        # For all InPorts that receive input from
+                        # virtual ports...
+                        transform_funcs = None
+                        if isinstance(pt, InPort):
+                            # ... extract a function pointer to the
+                            # transformation function of each virtual port.
+                            transform_funcs = \
+                                [vp.get_transform_func()
```
[the remark this answers is not in the document]
Comment (mathisrichter) [reply src/lava/magma/compiler/compiler.py:347]: I am raising an exception now and there is a unit test to check whether that exception is in fact raised in this case. Will have to solve this problem later.

File: src/lava/magma/core/process/ports/ports.py [hunk 4]
```diff
@@ -159,10 +159,17 @@ def get_incoming_virtual_ports(self) -> ty.List["AbstractVirtualPort"]:
...
             virtual_ports = []
+            num_virtual_ports = 0
             for p in self.in_connections:
                 virtual_ports += p.get_incoming_virtual_ports()
                 if isinstance(p, AbstractVirtualPort):
                     virtual_ports.append(p)
+                    num_virtual_ports += 1
+
+            if num_virtual_ports > 1:
+                raise NotImplementedError("joining multiple virtual ports is "
```
Comment (awintel) [inline src/lava/magma/core/process/ports/ports.py:170]: Joining...

**Review by awintel — APPROVED** [review]
Looks fine.

**Review by bamsumit — APPROVED** [review]
Looks good overall.

**Comment by mathisrichter** [comment]
Merging this in its current form for the release. This PR only supports virtual ports for reshape, flatten, and transpose between an OutPort and an InPort. Hierarchical Processes as well as chains of virtual ports are supported as well. But joining and forking as well as connection to and from other types of Ports (RefPorts, VarPorts) are not yet supported.

**Pull request merged**